Adding new guides for developing and deploying an ML application in S… (#36700)
ray-project/ray: doc/source/serve/develop-and-deploy.md

```

This command will run the `translator_app` application and then block, streaming logs to the console. It can be killed with `Ctrl-C`, which will tear down the application.

We can now test our model over HTTP. It can be reached at the following URL by default:

```
http://127.0.0.1:8000/
```

We'll send a POST request with JSON data containing our English text. Here's a client script that requests a translation for "Hello world!":

```{literalinclude} ../serve/doc_code/develop_and_deploy/model_deployment_with_fastapi.py
:start-after: __client_function_start__
:end-before: __client_function_end__
:language: python
```

While a Ray Serve application is deployed, you can use the `serve status` CLI command to check the status of the applications and deployments. For more details on the output format of `serve status`, please look at [Inspecting Serve in production guide](serve-in-production-inspecting).

```console
$ serve status
name: default
app_status:
  status: RUNNING
  message: ''
  deployment_timestamp: 1687415211.531879
deployment_statuses:
- name: default_Translator
  status: HEALTHY
  message: ''
```

## Building Serve config files for production deployment

> 

Reviewer: Suggest a sentence or two about why I need to create a Serve config file.

Author: @edoakes any suggestions for how to phrase the reasons?

Reviewer 2: "A Serve config file is the single source of truth for the cluster, allowing you to specify system-level configuration and your applications in one place. It also allows you to declaratively update your applications."

> In order to deploy serve applications in production, you need to first generate a serve config YAML file. This can be done using the `serve build` CLI command that takes as input the import path and can be saved to an output file using the `-o` flag. All deployment properties can be specified in the serve config files.

Reviewer: Suggestion explaining what the import path is to. Is it a path to a YAML file?

> In order to deploy serve applications in production, you need to first generate a serve config YAML file. This can be done using the `serve build` CLI command that takes as input the import path and can be saved to an output file using the `-o` flag. All deployment properties can be specified in the serve config files.

Author: Do you think its apparent based on the serve run instructions above? It's the same value as used in serve run

```console
$ serve build model:translator_app -o config.yaml
```

> This is what the serve config file will look like:

Reviewer: I don't follow where the settings for this config file came from.

> This is what the serve config file will look like:

Author: it comes from serve build

(serve-develop-and-deploy)=

# Develop and deploy an ML application

The flow for developing a Ray Serve application locally and deploying it in production covers the following steps:

* Converting a Machine Learning model into a Ray Serve application
* Testing the application locally
* Building Serve config files for production deployment

> * Deploying Ray Serve in production

Reviewer 2: ```suggestion
* Deploying applications using a config file
```

> ## Convert a Model into a Ray Serve application

Reviewer 2: ```suggestion
## Convert a model into a Ray Serve application
```

3. `bind` the `Translator` deployment to the arguments that are passed into its constructor

For other HTTP options, see [Set Up FastAPI and HTTP](serve-set-up-fastapi-http). 

```{literalinclude} ../serve/doc_code/develop_and_deploy/model_deployment_with_fastapi.py
:start-after: __deployment_start__
:end-before: __deployment_end__
:language: python
```

Note that the code configures parameters for the deployment, such as `num_replicas` and `ray_actor_options`. These parameters help configure the number of copies of the deployment and the resource requirements for each copy. In this case, we set up 2 replicas of the model that take 0.2 CPUs and 0 GPUs each. For a complete guide on the configurable parameters on a deployment, see [Configure a Serve deployment](serve-configure-deployment).

## Test a Ray Serve application locally

To test locally, run the script with the `serve run` CLI command. This command takes in an import path formatted as `module:application`. Run the command from a directory containing a local copy of the script saved as `model.py`, so it can import the application:

```console
$ serve run model:translator_app
```

> This command runs the `translator_app` application and then blocks streaming logs to the console. It can be killed with `Ctrl-C` kills it and tears down the application.

Reviewer: ```suggestion
This command runs the `translator_app` application and then blocks streaming logs to the console. You can kill it with `Ctrl-C`, which tears down the application.
```

Now test the model over HTTP. Reach it at the following default URL:

```
http://127.0.0.1:8000/
```

Send a POST request with JSON data containing the English text. This client script requests a translation for "Hello world!":

```{literalinclude} ../serve/doc_code/develop_and_deploy/model_deployment_with_fastapi.py
:start-after: __client_function_start__
:end-before: __client_function_end__
:language: python
```

While a Ray Serve application is deployed, use the `serve status` CLI command to check the status of the application and deployment. For more details on the output format of `serve status`, see [Inspect Serve in production](serve-in-production-inspecting).

```console
$ serve status
name: default
app_status:
  status: RUNNING
  message: ''
  deployment_timestamp: 1687415211.531879
deployment_statuses:
- name: default_Translator
  status: HEALTHY
  message: ''
```

## Build Serve config files for production deployment

> To deploy serve applications in production, you need to generate a Serve config YAML file. A Serve config file is the single source of truth for the cluster, allowing you to specify system-level configuration and your applications in one place. It also allows you to declaratively update your applications. The `serve build` CLI command takes as input the import path and saves to an output file using the `-o` flag. You can specify all deployment parameters in the Serve config files.

Reviewer: ```suggestion
To deploy Serve applications in production, you need to generate a Serve config YAML file. A Serve config file is the single source of truth for the cluster, allowing you to specify system-level configuration and your applications in one place. It also allows you to declaratively update your applications. The `serve build` CLI command takes as input the import path and saves to an output file using the `-o` flag. You can specify all deployment parameters in the Serve config files.
```

applications:

- name: app1

  route_prefix: /

  import_path: model:translator_app

  runtime_env: {}

  deployments:

  - name: Translator
    num_replicas: 2
    ray_actor_options:
      num_cpus: 0.2
      num_gpus: 0.0
```

You can also use the Serve config file with `serve run` for local testing. For example:

```console
$ serve run config.yaml
```

```console
$ serve status
name: app1
app_status:
  status: RUNNING
  message: ''
  deployment_timestamp: 1687630567.0700073
deployment_statuses:
- name: app1_Translator
  status: HEALTHY
  message: ''
```

For more details, see [Serve Config Files](serve-in-production-config-file).

## Deploy Ray Serve in production

Deploy the Ray Serve application in production on Kubernetes using the [KubeRay] operator. Copy the YAML file generated in the previous step directly into the Kubernetes configuration. KubeRay supports zero-downtime upgrades, status reporting, and fault tolerance for your production application. See [Deploying on Kubernetes](serve-in-production-kubernetes) for more information. For production usage, consider implementing the recommended practice of setting up [head node fault tolerance](serve-e2e-ft-guide-gcs).

> ## Monitor a Ray Serve cluster

Reviewer 2: ```suggestion
## Monitor Ray Serve
```

Reviewer 2: to match above
